subsys/bluetooth/host/gatt.c

```diff
@@ -5541,7 +5541,11 @@ int bt_gatt_subscribe(struct bt_conn *conn,
 	 */
 	sys_slist_prepend(&sub->list, &params->node);
 
-	return 0;
+	/*
+	 * Report already if subscription exist so that application does not wait
+	 * for the subscribe callback
+	 */
+	return has_subscription ? -EALREADY : 0;
```

Comment:
I'm not 100% certain this is the right approach. 

In this case we may return an error, while still adding the params to the list. That means this is a case where we return an error while still changing the state, which I'm not a fan off. 

Did you consider the approach of calling the callback instead without the GATT write when it exists? (Although I'm not a fan of that approach either tbh).

Reply:
I see your dilemma, it sounds wrong to change the state and still return an error.

I don't have a problem with that approach either. 
The disadvantage in that case is that we will be indicating to the application that a ccc write happened. I am not sure if this can have any implication to some application.

Reply:
Yeah, my assumption is that it is also a optimization.
I'll give it a try, but my assumption is that it will work without a problem.